### ISSUE TITLE
Can't redefine self.[] in a subclass of array

### ISSUE THREAD

**Issue by catmando** [opal/opal#842]
```
class S < Array
  def self.[](*args)
    puts "i never get to this line"
  end
end

S[1, 2, 3]
```
removing subclass of array and all works well

changing name from self.[] to self.build and all works well

**Comment by meh** [comment]
Subclasses of natively bridged classes are very very hacky right now, and they're hard to fix.

**Comment by vais** [comment]
@catmando like @meh said, `String` and `Array` inheritance only kinda sorta works in Opal.

I was pretty bummed out about it at first, but then I remembered that I have never actually subclassed `String` or `Array` in my "normal" Ruby usage, and that **it is generally not recommended in Ruby to subclass core classes** (see [this excellent explanation by Steve Klabnik](http://words.steveklabnik.com/beware-subclassing-ruby-core-classes)).

So, subclassing core Ruby classes is a bad idea even in MRI where it obviously *does* work. In the end, the way I coped with this presently *not* working in Opal is I decided that I never liked it *anyway* :smile:

**Comment by vais** [comment]
The more I think about it, @meh, http://words.steveklabnik.com/beware-subclassing-ruby-core-classes gives some very good reasons for pulling native inheritance out of Opal altogether. This way you can forget the native bridging rewrite as a nightmare. Or is that going too far?

cc @adambeynon @elia

**Comment by Ajedi32** [comment]
I don't think it's a good idea to break compatibility with MRI except in places where we absolutely have to. Ideally we would be able to pass any pure-Ruby library or program into Opal and get working JavaScript code out the other end.

It's fine if you want to prioritize other features over subclassing of core classes (as I agree that that kind  of subclassing isn't something people should be doing under most circumstances), but I think for compatibility with existing programs the ability to subclass core classes is definitely useful.

**Comment by meh** [comment]
Yeah, we're not going to disable it completely, but it is a low priority, especially since it's quite hard to do right.

**Comment by elia** [comment]
fixed by #1206
